Tensor reinitialization codemod - 3/5

jerryzh168 opened the pull request:
Summary:
Codemod generated with clangr shard mode, 25 files per diff,
To eliminiate partially initialized Tensor, we split the initialization of local Tensor variables into two steps, first declare un uninitialized Tensor, and
call `ReinitializeTensor` to initialize it.
motivation: https://github.com/pytorch/pytorch/pull/12407

Differential Revision: D13586734

smessmer commented on caffe2/operators/experimental/c10/schemas/fc.h:
```diff
@@ -12,6 +12,7 @@ struct FullyConnected final {
 
   struct Cache final {
     vector<int64_t> Y_shape_cache_;
+    // TODO: Remove Tensor{CPU}
```
This bias_multiplier_ tensor is only used in caffe2/operators/experimental/c10/cpu/fc_cpu.cpp. That's where your Reinitialize call should be.

jerryzh168 replied:
But now it's a C10Tensor isn't it? do we have to write a ReinitializeTensor for C10Tensor as well?

smessmer replied:
C10Tensor is just a thin wrapper around TensorImpl. It doesn't do anything different internally. If the kernel expects the tensor to be initialized the caffe2 way (which fc probably does), it has to be initialized the caffe2 way.

ezyang commented on caffe2/operators/distance_op.cu:
```diff
@@ -316,7 +316,7 @@ bool CosineSimilarityOp<float, CUDAContext>::RunOnDevice() {
   const float* Y_data = Y.data<float>();
   // Auxiliary arrays, one allocation of memory
-  aux_.Resize(2 * N);
+  ReinitializeTensor(&aux_, {2 * N}, at::dtype<float>().device(CUDA));
```
Isn't the specification of CUDA here supposed to be unnecessary, because we should infer the device type from the Operator context?

jerryzh168 replied:
ReinitializeTensor is a free function, we can use it outside of Operator class as well. Also we can have Tensor{CPU} even when context is CUDA, so it's not necessarily equivalent to the Context.
This is a natural codemod from the declaration of "Tensor x{CUDA};" We can have a special wrapper for it inside Operator though.

jerryzh168 committed:
Tensor reinitialization codemod - 3/5 (#15912)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15912

Codemod generated with clangr shard mode, 25 files per diff,
To eliminiate partially initialized Tensor, we split the initialization of local Tensor variables into two steps, first declare un uninitialized Tensor, and
call `ReinitializeTensor` to initialize it.
motivation: https://github.com/pytorch/pytorch/pull/12407

Reviewed By: dzhulgakov

Differential Revision: D13586734

fbshipit-source-id: 574b7d90b5e59136139fe7083c9ae54393efccc4
